docs(tutorials): Add SDB x Containers x Strapi tutorial MTA-4593 (#3182)
scaleway/docs-content: tutorials/strapi-app-serverless-containers-sqldb/index.mdx

5. Run the following command to display them:
    ```
    terraform output -json
    ```

    A similar output displays:

    ```json
    {
     "container_url": {
       "sensitive": true,
       "type": "string",
       "value": "tutorialstrapiblogtfaxtypxrf-tutorial-strapi-blog-tf.functions.fnc.fr-par.scw.cloud"
     },
     "database_connection_string": {
       "sensitive": true,
       "type": "string",
       "value": "postgres://example-f6b7-40cc-9ae8-7f24e64c6531:example-c770-46ea-b785-94bf39536e6a@650c9680-1100-48e4-b5a6-ff2ff5bcf142.pg.sdb.fr-par.scw.cloud:5432/tutorial-strapi-tf?sslmode=require"
     }
    }
    ```

6. Copy the container URL from the output and paste it into your browser. The main Strapi page displays. Click on "Open the administration" or add `/admin` to your browser URL to access Strapi Administration Panel.

    <Message type="tip">
    The first deployment can take a few minutes. You can check the deployment status with the following command:
    ```sh
    scw container container list name=tutorial-strapi-blog-tf
    ```
    When the status appears as `ready`, you can access the Strapi Administration Panel via your browser.
    </Message>

> 7. (Optional) You can check that Strapi APIs are working with the following command or going to `https://{container_url}:1337/api/articles` in your browser:

Author: This step returns the following error
```
curl: (7) Failed to connect to tutorialstrapiblogtf0hm5uzfe-tutorial-strapi-blog-tf.functions.fnc.fr-par.scw.cloud port 1337 after 8 ms: Couldn't connect to server
```

---
meta:
  title: Deploy a Strapi headless CMS using Serverless Containers and Serverless SQL Databases
  description: This page explains how to deploy a fully serverless Strapi headless CMS using Serverless Containers and Serverless SQL Databases
content:
  h1: Deploy a Strapi headless CMS using Serverless Containers and Serverless SQL Databases
  paragraph: This page explains how to deploy a fully serverless Strapi headless CMS using Serverless Containers and Serverless SQL Databases
dates:
  validation: 2024-05-21
  posted: 2024-05-21
---

This tutorial will guide you through deploying a fully serverless Strapi headless CMS using a Serverless Container and a Serverless SQL Database.

You can either deploy your application:

>   - [step by step using Scaleway CLI](#deploy-strapi-headless-cms-using-scaleway-cli) to understand each detailed action performed, and the resources and subresources required

Reviewer: ```suggestion
  - [step by step using the Scaleway CLI](#deploy-strapi-headless-cms-using-scaleway-cli) to understand each detailed action performed, and the resources and subresources required
```

>   - [step by step using the Scaleway CLI](#deploy-strapi-headless-cms-using-scaleway-cli) to understand each detailed action performed, and the resources and subresources required

Reviewer 2: ```suggestion
  - [step by step using the Scaleway CLI](#deploy-strapi-headless-cms-using-scaleway-cli) to understand each action performed in details, and the resources required
```

>   - [using a Terraform template](#deploy-strapi-using-scaleway-terraform-templates) to deploy your application faster and have a ready-to-use Infrastructure as Code.

Reviewer: ```suggestion
  - [using a Terraform template](#deploy-strapi-using-scaleway-terraform-templates) to deploy your application faster and have ready-to-use Infrastructure as Code.
```

<Macro id="requirements" />

- A Scaleway account logged into the [console](https://console.scaleway.com)
- [Owner](/identity-and-access-management/iam/concepts/#owner) status or [IAM permissions](/identity-and-access-management/iam/concepts/#permission) allowing you to perform actions in the intended Organization
- A valid [API key](/identity-and-access-management/iam/how-to/create-api-keys/)

> - Installed and initialized the [Scaleway CLI](https://www.scaleway.com/en/docs/developer-tools/scaleway-cli/quickstart/)

Reviewer: ```suggestion
- Installed and initialized the [Scaleway CLI](/developer-tools/scaleway-cli/quickstart/)
```

    --dbport=$DATABASE_PORT --dbname=$DATABASE_NAME \
    --dbusername=$DATABASE_USERNAME \
    --dbpassword=$DATABASE_PASSWORD --dbssl=true 
    ```

2. Access the folder you just created:
    ```bash
    cd my-blog
    ```

3. Start Strapi in development mode:
    ```bash
    npm run develop
    ```

    Your browser should open automatically at the address [http://localhost:1337](http://localhost:1337) to create a local administrator account.
    Define your credentials and create the account.

4. You can now access Strapi Administration Panel at [http://localhost:1337/admin](http://localhost:1337/admin). Once connected with your local administrator account, you can browse and edit content.

### Building the Strapi container

1. Run the following command to create a `Dockerfile`:
    ```bash
    touch Dockerfile
    ```
2. Add the following code to it:
    ```bash
    # Creating a multi-stage build for production
    FROM node:20-alpine as build
    RUN apk update && apk add --no-cache build-base gcc autoconf automake zlib-dev libpng-dev vips-dev git > /dev/null 2>&1
    ARG NODE_ENV=production
    ENV NODE_ENV=${NODE_ENV}

    WORKDIR /opt/

>     #Copy package-lock.json only if it exists

Reviewer: ```suggestion
    # Copy package-lock.json only if it exists
```

    </Message>
  
3. Copy the endpoint URL displayed next to the `DomainName` property, and paste it into your browser. The main Strapi page displays. Click on "Open the administration" or add `/admin` to your browser URL to access Strapi Administration Panel.

4. (Optional) You can check that Strapi APIs are working with the following command, or  by accessing `https://{container_url}/api/articles` in your browser:
    ```bash
    curl -X GET https://{container_url}/api/articles | jq
    ```

    The available articles should display in JSON format:
    ```json
    {
      "data": [
      {
        "id": 1,
        "attributes": {
          "title": "What's inside a Black Hole",
          "description": "Maybe the answer is in this article, or not...",
          "slug": "what-s-inside-a-black-hole",
          "createdAt": "2024-04-16T15:49:14.525Z",
          "updatedAt": "2024-04-16T15:49:14.525Z",
          "publishedAt": "2024-04-16T15:49:14.457Z"
        }
      },
        (...)
      ]
    }
    ```

Congratulations! Your Strapi headless CMS for a blog is now online. You can already edit stored content, or consume it from a frontend application.

> Moreover, your textual content data is already stored in serverless storage, so that your full application only runs when you connect to the administration panel or perform requests to Strapi API.

Reviewer 2: ```suggestion
Moreover, your textual content data is already stored in serverless storage, so that your full application only runs when you connect to the administration panel or perform requests to the Strapi API.
```

However, your Strapi container currently connects to your database with your [user credentials](/identity-and-access-management/iam/concepts/#user), which is not a recommended security practice.

## Secure and redeploy your application

To secure your deployment, we will now add a dedicated [IAM application](/identity-and-access-management/iam/concepts/#application), give it the minimum required permissions, and provide its credentials to your Strapi container.

1. Run the following command to create an [IAM application](/identity-and-access-management/iam/concepts/#application) and export it as a variable: 
    ```bash
    export SCW_APPLICATION_ID=$(scw iam application create name=tutorial-strapi-blog -o json | jq -r '.id')
    ```

>     The `SCW_APPLICATION_ID` environment variable will store the IAM application ID so you can use it in later commands.

Reviewer 2: ```suggestion
    The `SCW_APPLICATION_ID` environment variable will store the IAM application ID so you can use it in the next commands.
```

2. Create an [IAM policy](/identity-and-access-management/iam/concepts/#policy) giving your application rights to access the database:
    ```bash
    scw iam policy create application-id=$SCW_APPLICATION_ID name=tutorial-strapi-policy rules.0.organization-id=$(scw config get default-organization-id) rules.0.permission-set-names.0=ServerlessSQLDatabaseFullAccess
    ```

>     The permission `ServerlessSQLDatabaseFullAccess` lets your application read and write data or create new databases. You can restrict this later to fit your database use cases, using `ServerlessSQLDatabaseRead` or `ServerlessSQLDatabaseReadWrite` permissions sets.

Reviewer 2: ```suggestion
    The `ServerlessSQLDatabaseFullAccess` permission lets your application read and write data or create new databases. You can restrict this later to fit your database use cases, using the `ServerlessSQLDatabaseRead` or `ServerlessSQLDatabaseReadWrite` permissions sets.
```

3. Create an [IAM API Key](/identity-and-access-management/iam/concepts/#api-key) and export its secret key as an environment variable:
    ```bash
    export SCW_APPLICATION_SECRET=$(scw iam api-key create application-id=$SCW_APPLICATION_ID -o json | jq -r '.secret_key')
    ```

4. Retrieve your container's ID:
    ```bash
    export CONTAINER_ID=$(scw container container list name="my-strapi-blog" -o json | jq -r '.[0].id')
    ```

> 8. Redeploy your Serverless Container:

Reviewer 2: ```suggestion
5. Redeploy your Serverless Container:
```

    ```bash
    scw container container update $CONTAINER_ID \
    environment-variables.DATABASE_HOST=$DATABASE_HOST \
    environment-variables.DATABASE_PORT=$DATABASE_PORT \
    environment-variables.DATABASE_NAME=$DATABASE_NAME \
    environment-variables.DATABASE_CLIENT=$DATABASE_CLIENT \
    environment-variables.DATABASE_SSL=$DATABASE_SSL \
    environment-variables.DATABASE_USERNAME=$SCW_APPLICATION_ID \
    secret-environment-variables.0.key='DATABASE_PASSWORD' \
    secret-environment-variables.0.value=$SCW_APPLICATION_SECRET \
    secret-environment-variables.1.key='APP_KEYS' \
    secret-environment-variables.1.value=$APP_KEYS \
    secret-environment-variables.2.key='API_TOKEN_SALT' \
    secret-environment-variables.2.value=$API_TOKEN_SALT \
    secret-environment-variables.3.key='ADMIN_JWT_SECRET' \
    secret-environment-variables.3.value=$ADMIN_JWT_SECRET \
    secret-environment-variables.4.key='TRANSFER_TOKEN_SALT' \
    secret-environment-variables.4.value=$TRANSFER_TOKEN_SALT \
    secret-environment-variables.5.key='JWT_SECRET' \
    secret-environment-variables.5.value=$JWT_SECRET redeploy=true
    ```

> 9. Refresh your browser page displaying Strapi Administration Panel. An updated version displays. 

Reviewer 2: ```suggestion
6. Refresh your browser page displaying the Strapi Administration Panel. An updated version displays. 
```

>   Congratulations, you have deployed a full serverless Strapi blog example!

Reviewer 2: ```suggestion
  You have now deployed a full serverless Strapi blog example!
```

   
## Going further with containers

- Inspect your newly created resources in the Scaleway console:

  - You can display your **Registry namespace** and **container image** in the [Container Registry section](https://console.scaleway.com/registry/namespaces).

  - You can display your **Serverless Containers namespace** and **container deployment** in the [Serverless Containers section](https://console.scaleway.com/containers/namespaces).

  - You can display your **Serverless SQL Database** in the [Serverless SQL Databases section](https://console.scaleway.com/serverless-db/databases).

> - **Fine tune deployment options** such as autoscaling, targeted regions, and more. You can find more information by typing `scw container deploy --help` in your terminal, or by referring to the [dedicated documentation](/serverless/containers/how-to/manage-a-container/)

Reviewer 2: ```suggestion
- **Fine tune deployment options** such as autoscaling, targeted regions, and more. You can find more information by typing `scw container deploy --help` in your terminal, or by referring to the [dedicated documentation](/serverless/containers/how-to/manage-a-container/).
```

> - **Fine-tune deployment options** such as autoscaling, targeted regions, and more. You can find more information by typing `scw container deploy --help` in your terminal, or by referring to the [dedicated documentation](/serverless/containers/how-to/manage-a-container/)

Reviewer: ```suggestion
- **Fine tune deployment options** such as autoscaling, targeted regions, and more. You can find more information by typing `scw container deploy --help` in your terminal, or by referring to the [dedicated documentation](/serverless/containers/how-to/manage-a-container/)
```

> - Create a secondary development environment by duplicating your built container, building it in `NODE_ENV=development` environment, running `npm run develop`, and plugging it onto another **Serverless SQL Database**. This will let you for instance edit content types which is not possible in production.

Reviewer: ```suggestion
- Create a secondary development environment by duplicating your built container, building it in `NODE_ENV=development` environment, running `npm run develop`, and plugging it onto another **Serverless SQL Database**. This will let you, for instance, edit content types, which is not possible in production.
```

<Message type="note">
  You must have [Terraform](https://developer.hashicorp.com/terraform/tutorials/aws-get-started/install-cli) installed and configured with the [Scaleway Terraform Provider](https://registry.terraform.io/providers/scaleway/scaleway/latest/docs) for this part.
</Message>

### Initialize the project

1. Run the command below to export your access key and secret key as environment variables:
    ```sh
    export SCW_ACCESS_KEY=$(scw config get access-key)
    export SCW_SECRET_KEY=$(scw config get secret-key)
    ```

2. Run the command below to make sure the environment variables are properly set:

    ```sh
    scw info 
    ```

    This command displays your access_key and secret_key in the two last lines of the output. The `ORIGIN` column should display `env (SCW_ACCESS_KEY)` and `env (SCW_SECRET_KEY)`, and not `default profile`.
    
    ```bash
    KEY                      VALUE                                       ORIGIN
    (...)
    access_key               SCWF9DETBF829TAFB1TB                        env (SCW_ACCESS_KEY)
    secret_key               9a1bcf92-xxxx-xxxx-xxxx-xxxxxxxxxxxx        env (SCW_SECRET_KEY)
    ```

> ### Building Strapi container

Reviewer 2: ```suggestion
### Building a Strapi container
```

    CMD npm run --silent strapi admin:create-user -- --firstname=$ADMIN_FIRSTNAME --lastname=$ADMIN_LASTNAME --email=$ADMIN_EMAIL --password=$ADMIN_PASSWORD ; (npm run --silent strapi admin:reset-user-password -- --email=$ADMIN_EMAIL --password=$ADMIN_PASSWORD && npm run develop)
    ```

6. Create a `.dockerignore` file:
    ```bash
    touch .dockerignore
    ```

7. Add the following code to it:
    ```bash
    .tmp/
    .cache/
    .git/
    .env
    build/
    node_modules/
    backup/
    ```

    Docker will now ignore node dependencies and other unnecessary files during the build process.

    At this stage, your application folder should be called `my-blog` and structured as follows:
      ```bash
      my-blog/
      ├── build/
      ├── config/
      ├── data/
      ├── database/
      ├── node_modules/
      ├── public/
      ├── src/
      ├── types/
      ├── .dockerignore
      ├── Dockerfile
      ├── favicon.png
      ├── jsconfig.json
      ├── package.json
      ├── README.md
      └── yarn.lock					
      ```

8. Build your application container:
    ```bash
    docker build -t my-strapi-blog .   
    ```
    <Message type="tip">

>     The docker build image process can take a few minutes, particularly during `npm install` step since Strapi requires around 1 GB of node modules to be built.

Reviewer 2: ```suggestion
    The docker build image process can take a few minutes, particularly during the `npm install` step since Strapi requires around 1 GB of node modules to be built.
```

1. Run the following command to create a [Container Registry namespace](/containers/container-registry/concepts/#namespace) and export its endpoint as a variable:
    ```bash
    export REGISTRY_ENDPOINT=$(scw registry namespace create -o json | jq -r '.endpoint')
    ```

2. Log in to your [Container Registry](/containers/container-registry/concepts/#container-registry) from your local terminal:
    ```bash
    docker login $REGISTRY_ENDPOINT -u nologin --password-stdin <<< "$SCW_SECRET_KEY"
    ```
    The following output displays:
    ```sh
    Login Succeeded
    ```

3. Tag and push your container image to your Container Registry namespace:
    ```bash
    docker tag my-strapi-blog:latest $REGISTRY_ENDPOINT/my-strapi-blog:latest
    docker push $REGISTRY_ENDPOINT/my-strapi-blog:latest
    ```

  
### Creating the Terraform configuration

1. Run the following command to create a new folder to store your Terraform files, and access it:
    ```bash
    cd ..
    mkdir terraform-strapi-blog &&
    cd terraform-strapi-blog
    ``` 

2. Create an empty `main.tf` Terraform file inside the folder.

    Your application folder should now be structured as follows:

>     ```bash
>     my-blog/
>     terraform-strapi-blog/
>         main.tf
>     ```

Reviewer 3: ```suggestion
    ```bash
    my-blog/
    └── terraform-strapi-blog/
        └── main.tf
    ```
```

            permission_set_names = ["ServerlessSQLDatabaseReadWrite"]
        }
    }
    
    resource scaleway_iam_api_key "api_key" {
        application_id = scaleway_iam_application.app.id
    }
    
    resource scaleway_sdb_sql_database "database" {
        name = "tutorial-strapi-tf"
        min_cpu = 0
        max_cpu = 8
    }
    
    output "database_connection_string" {
        // Output as an example, you can give this string to your application
        value = format("postgres://%s:%s@%s",
            scaleway_iam_application.app.id,
            scaleway_iam_api_key.api_key.secret_key,
            trimprefix(scaleway_sdb_sql_database.database.endpoint, "postgres://"),
        )
        sensitive = true
    }
    
    output "container_url" {
        // Output as an example, you can give this string to your application
        value = scaleway_container.main.domain_name
        sensitive = true
    }
    ```

The Terraform file creates several resources:

- A [Serverless Containers namespace](/serverless/containers/concepts/#namespace), that contains a [Serverless Container](/serverless/containers/quickstart/), which hosts your Strapi application

> - A [Serverless SQL Database](/serverless/sql-databases/reference-content/serverless-sql-databases-overview/) which stores posts data

Reviewer 2: I don't understand what you mean by "posts data", is it data from blog posts?

Author: yes !

- An [IAM application](/identity-and-access-management/iam/concepts/#application), used as a principal for your IAM policy

- An [API key](/identity-and-access-management/iam/concepts/#api-key) used as credentials to authenticate your application to the database

### Deploying the app with Terraform

1. Run the following command to initialize the Terraform working directory:
    ```bash
    terraform init
    ```

2. Add the `REGISTRY_ENDPOINT`, `DEFAULT_PROJECT_ID`, `ADMIN_EMAIL` and `ADMIN_PASSWORD` environment variables to Terraform:
    ```bash
    export TF_VAR_REGISTRY_ENDPOINT=$REGISTRY_ENDPOINT
    export TF_VAR_DEFAULT_PROJECT_ID=$(scw config get default-project-id)
    export TF_VAR_ADMIN_EMAIL=admin@example.com
    export TF_VAR_ADMIN_PASSWORD=Y0ur_P@ssw0rd
    ```
    Edit the `ADMIN_EMAIL` and `ADMIN_PASSWORD` values with your own email and password. Optionally, you can also edit `ADMIN_FIRSTNAME` and `ADMIN_LASTNAME` values to change the default admin first and last name.
    <Message type="note">
        Strapi admin password requires at least 8 characters including one uppercase, one lowercase, one number, and one special character. 

>         If the admin password or email does not meet the requirements, the container will not start.

Reviewer 2: ```suggestion
        If the admin password or email does not meet the requirements, the container will not start.
```

    ```
    terraform output -json
    ```

    A similar output displays:

    ```json
    {
     "container_url": {
       "sensitive": true,
       "type": "string",
       "value": "tutorialstrapiblogtfaxtypxrf-tutorial-strapi-blog-tf.functions.fnc.fr-par.scw.cloud"
     },
     "database_connection_string": {
       "sensitive": true,
       "type": "string",
       "value": "postgres://example-f6b7-40cc-9ae8-7f24e64c6531:example-c770-46ea-b785-94bf39536e6a@650c9680-1100-48e4-b5a6-ff2ff5bcf142.pg.sdb.fr-par.scw.cloud:5432/tutorial-strapi-tf?sslmode=require"
     }
    }
    ```

6. Copy the container URL from the output and paste it into your browser. The main Strapi page displays. Click on "Open the administration" or add `/admin` to the container URL to access the Strapi Administration Panel.

    <Message type="tip">
    The first deployment can take a few minutes. You can check the deployment status with the following command:
    ```sh
    scw container container list name=tutorial-strapi-blog-tf -o human=Name,Status
    ```
    When the status appears as `ready`, you can access the Strapi Administration Panel via your browser.
    </Message>

> 7. (Optional) You can check that Strapi APIs are working with the following command or going to `https://{container_url}/api/articles` in your browser:

Reviewer: ```suggestion
7. (Optional) You can check that Strapi APIs are working with the following command, or by going to `https://{container_url}/api/articles` in your browser:
```

    ```bash
    curl -X GET https://{container_url}/api/articles | jq
    ```

    The available articles display in JSON format:
    ```json
    {
      "data": [
      {
        "id": 1,
        "attributes": {
          "title": "What's inside a Black Hole",
          "description": "Maybe the answer is in this article, or not...",
          "slug": "what-s-inside-a-black-hole",
          "createdAt": "2024-04-16T15:49:14.525Z",
          "updatedAt": "2024-04-16T15:49:14.525Z",
          "publishedAt": "2024-04-16T15:49:14.457Z"
        }
      },
        (...)
      ]
    }
    ```

> Congratulations! Your Strapi headless CMS for a blog is now online. You can already edit stored content, or consume it from a frontend application.

Reviewer 2: ```suggestion
Your Strapi headless CMS for a blog is now online. You can already edit stored content, or consume it from a frontend application.
```

  - You can display your **Registry namespace** and **container image** in the [Container Registry section](https://console.scaleway.com/registry/namespaces)

  - You can display your **Serverless Containers namespace** and **container deployment** in the [Serverless Containers section](https://console.scaleway.com/containers/namespaces)

  - You can display your **Serverless SQL Database** in the [Serverless SQL Databases section](https://console.scaleway.com/serverless-db/databases)

- **Fine-tune deployment options** such as autoscaling, targeted regions, and more. You can find more information by typing `scw container deploy --help` in your terminal, or by referring to the [dedicated documentation](/serverless/containers/how-to/manage-a-container/)

- Create a secondary production environment by duplicating your built container, building it in `NODE_ENV=production` environment, running `npm run start`, and plugging it onto another **Serverless SQL Database**. For instance, this will allow you to edit content-types which is not possible in production.
 
## Troubleshooting

If you happen to encounter any issues, first check that you meet all the requirements.

> - You have installed and configured the [Scaleway CLI](https://www.scaleway.com/en/docs/developer-tools/scaleway-cli/quickstart/). Running the `scw account project get` command should return the following output:

Reviewer 2: ```suggestion
- You have installed and configured the [Scaleway CLI](/developer-tools/scaleway-cli/quickstart/). Running the `scw account project get` command should return the following output:
```

    ```
    ID              example-7afa-4a5d-9f6c-27db072f1527
    Name            default
    OrganizationID  example-7afa-4a5d-9f6c-27db072f1527
    CreatedAt       1 year ago
    UpdatedAt       1 year ago
    Description     -
    ```
  <Message type="tip">  

>   You can also find and compare your Project and organization ID in the [Scaleway console settings](https://console.scaleway.com/project/settings).

Reviewer: ```suggestion
  You can also find and compare your Project and Organization ID in the [Scaleway console settings](https://console.scaleway.com/project/settings).
```
